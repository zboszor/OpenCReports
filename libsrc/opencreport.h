/*
 * OpenCReports main header
 * Copyright (C) 2019 Zoltán Böszörményi <zboszor@gmail.com>
 * See COPYING.LGPLv3 in the toplevel directory.
 */
#ifndef _OPENCREPORT_H_
#define _OPENCREPORT_H_

#include <stdbool.h>
#include <stdint.h>
#include <time.h>
#include <mpfr.h>

/* Main report structure type */
struct opencreport;
typedef struct opencreport opencreport;

struct ocrpt_expr;
typedef struct ocrpt_expr ocrpt_expr;

enum ocrpt_result_type {
	/*
	 * Error in parsing or evaluation
	 */
	OCRPT_RESULT_ERROR,

	/*
	 * Constants
	 */
	OCRPT_RESULT_STRING,
	OCRPT_RESULT_NUMBER,
	OCRPT_RESULT_DATETIME
};

struct ocrpt_string {
	char *str;
	size_t allocated_len;
	size_t len;
};
typedef struct ocrpt_string ocrpt_string;

struct ocrpt_result {
	enum ocrpt_result_type type;
	/* Original lexer token or (computed) string value for expression */
	ocrpt_string *string;
	/* Converted numeric constant or computed numeric value for expression */
	mpfr_t number;
	bool number_initialized;
	bool string_owned;
	bool isnull;
	/* Datetime value */
	struct tm datetime;
};
typedef struct ocrpt_result ocrpt_result;

struct ocrpt_query;
typedef struct ocrpt_query ocrpt_query;

struct ocrpt_query_result {
	const char *name;
	bool name_allocated;
	ocrpt_result result;
};
typedef struct ocrpt_query_result ocrpt_query_result;

struct ocrpt_input {
	void (*describe)(ocrpt_query *, ocrpt_query_result **, int32_t *);
	void (*rewind)(ocrpt_query *);
	bool (*next)(ocrpt_query *);
	bool (*populate_result)(ocrpt_query *);
	bool (*isdone)(ocrpt_query *);
	void (*free)(ocrpt_query *);
};

typedef struct ocrpt_input ocrpt_input;

struct ocrpt_datasource;
typedef struct ocrpt_datasource ocrpt_datasource;

struct ocrpt_paper {
	const char *name;
	double width;
	double height;
};
typedef struct ocrpt_paper ocrpt_paper;

#if 0
struct report_line;
typedef struct report_line report_line;

struct ocrpt_report_part;
typedef struct ocrpt_report_part ocrpt_report_part;
#endif

#define OCRPT_MPFR_PRECISION_BITS	(256)

/*
 * Create a new empty OpenCReports structure
 */
opencreport *ocrpt_init(void);
/*
 * Free an OpenCReports structure
 */
void ocrpt_free(opencreport *o);
/*
 * Set MPFR numeric precision
 */
void ocrpt_set_numeric_precision_bits(opencreport *o, mpfr_prec_t prec);
/*
 * Set MPFR rounding mode
 */
void ocrpt_set_rounding_mode(opencreport *o, mpfr_rnd_t rndmode);

/********************************
 * Expression related functions *
 ********************************/

/*
 * Create an expression parse tree from an expression string
 */
ocrpt_expr *ocrpt_expr_parse(opencreport *o, const char *str, char **err);
/*
 * Optimize expression after parsing
 */
void ocrpt_expr_optimize(opencreport *o, ocrpt_expr *e);
/*
 * Resolve variable references in the expression
 */
void ocrpt_expr_resolve(opencreport *o, ocrpt_expr *e);
/*
 * Evaluate the expression, i.e. compute its ocrpt_result
 * It must be called after ocrpt_navigate_next(), see below.
 */
ocrpt_result *ocrpt_expr_eval(opencreport *o, ocrpt_expr *e);
/*
 * Print an expression on stdout. Good for unit testing.
 */
void ocrpt_expr_print(opencreport *o, ocrpt_expr *e);
/*
 * Print the result data. Good for unit testing.
 */
void ocrpt_expr_result_print(ocrpt_result *r);
/*
 * Count the number of expression nodes
 */
int ocrpt_expr_nodes(ocrpt_expr *e);
/*
 * Free an expression parse tree
 */
void ocrpt_free_expr(ocrpt_expr *e);

/****************************
 * Memory handling wrappers *
 ****************************/

typedef void *(*ocrpt_mem_malloc_t)(size_t);
typedef void *(*ocrpt_mem_realloc_t)(void *, size_t);
typedef void (*ocrpt_mem_free_t)(const void *);
typedef char *(*ocrpt_mem_strdup_t)(const char *);
typedef char *(*ocrpt_mem_strndup_t)(const char *, size_t);

extern ocrpt_mem_malloc_t ocrpt_mem_malloc0;
extern ocrpt_mem_realloc_t ocrpt_mem_realloc0;
extern ocrpt_mem_free_t ocrpt_mem_free0;
extern ocrpt_mem_strdup_t ocrpt_mem_strdup0;
extern ocrpt_mem_strndup_t ocrpt_mem_strndup0;

static inline void *ocrpt_mem_malloc(size_t sz) __attribute__((malloc,alloc_size(1)));
static inline void *ocrpt_mem_malloc(size_t sz) { return ocrpt_mem_malloc0(sz); }

static inline void *ocrpt_mem_realloc(void *ptr, size_t sz) __attribute__((alloc_size(2)));
static inline void *ocrpt_mem_realloc(void *ptr, size_t sz) { return ocrpt_mem_realloc0(ptr, sz); }

static inline void ocrpt_mem_free(const void *ptr) { ocrpt_mem_free0(ptr); }
static inline void ocrpt_strfree(const char *s) { ocrpt_mem_free0((const void *)s); }

static inline void *ocrpt_mem_strdup(const char *ptr) { return (ptr ? ocrpt_mem_strdup0(ptr) : NULL); }

static inline void *ocrpt_mem_strndup(const char *ptr, size_t sz) { return (ptr ? ocrpt_mem_strndup0(ptr, sz) : NULL); }

void ocrpt_mem_set_alloc_funcs(ocrpt_mem_malloc_t rmalloc,
								ocrpt_mem_realloc_t rrealloc,
								ocrpt_mem_free_t rfree,
								ocrpt_mem_strdup_t rstrdup,
								ocrpt_mem_strndup_t rstrndup);

/*
 * String related functions
 */
ocrpt_string *ocrpt_mem_string_new(const char *str, bool copy);
ocrpt_string *ocrpt_mem_string_new_with_len(const char *str, size_t len);
char *ocrpt_mem_string_free(ocrpt_string *string, bool free_str);
void ocrpt_mem_string_append_len(ocrpt_string *string, const char *str, const size_t len);
void ocrpt_mem_string_append(ocrpt_string *string, const char *str);
void ocrpt_mem_string_append_c(ocrpt_string *string, const char c);

/********************************
 * Paper size related functions *
 ********************************/

/*
 * System default paper name and size
 */
const ocrpt_paper *ocrpt_get_system_paper(void);
/*
 * Get paper size of the specified paper name
 */
const ocrpt_paper *ocrpt_get_paper_by_name(const char *paper);
/*
 * Set paper for the report using an ocrpt_paper structure
 * The contents of the structure is copied.
 */
void ocrpt_set_paper(opencreport *o, const ocrpt_paper *paper);
/*
 * Set paper for the report using a paper name
 * If the paper name is unknown, the system default paper is set.
 */
void ocrpt_set_paper_by_name(opencreport *o, const char *paper);
/*
 * Get the report's currently set paper
 */
const ocrpt_paper *ocrpt_get_paper(opencreport *o);
/*
 * Iterator for supported paper names and sizes
 * The returned paper structures are library-global
 * but the iterator is per report.
 */
const ocrpt_paper *ocrpt_paper_first(opencreport *o);
const ocrpt_paper *ocrpt_paper_next(opencreport *o);

/******************************************
 * Datasource and query related functions *
 ******************************************/

/*
 * Add a custom datasource not covered by
 * the currently implemented sources.
 */
ocrpt_datasource *ocrpt_add_datasource(opencreport *o, const char *source_name, const ocrpt_input *input);
/*
 * Find the datasource using its name
 */
ocrpt_datasource *ocrpt_find_datasource(opencreport *o, const char *source_name);
/*
 * Validate the datasource pointer against the report structure
 */
ocrpt_datasource *ocrpt_validate_datasource(opencreport *o, ocrpt_datasource *source);
/*
 * Add an array datasource
 *
 * Calling this is optional, as an array datasource called
 * "array" is automatically added to an opencreport structure.
 */
ocrpt_datasource *ocrpt_add_array_datasource(opencreport *o, const char *source_name);
/*
 * Add an array query using the datasource pointer
 */
ocrpt_query *ocrpt_add_array_query(opencreport *o, ocrpt_datasource *source,
									const char *name, void *array,
									int32_t rows, int32_t cols,
									const enum ocrpt_result_type *types);
/*
 * Add an array query using the datasource name
 */
ocrpt_query *ocrpt_add_array_query_as(opencreport *o, const char *source_name,
									const char *name, void *array,
									int32_t rows, int32_t cols,
									const enum ocrpt_result_type *types);
/*
 * Return the query result array and the number of columns in it
 *
 * It must be re-run for every new data source row since
 * the pointer is invalidated after ocrpt_navigate_next()
 */
ocrpt_query_result *ocrpt_query_get_result(ocrpt_query *q, int32_t *cols);
/*
 * Add follower query with a match function
 */
bool ocrpt_add_query_follower_n_to_1(opencreport *o,
									ocrpt_query *leader,
									ocrpt_query *follower,
									ocrpt_expr *match);
/*
 * Add follower query (runs side-by-side with "leader")
 */
bool ocrpt_add_query_follower(opencreport *o,
									ocrpt_query *leader,
									ocrpt_query *follower);
/*
 * Free a query and remove it from follower references
 */
void ocrpt_free_query(opencreport *o, ocrpt_query *q);
/*
 * Start query navigation from the beginning of the resultset
 */
void ocrpt_navigate_start(opencreport *o, ocrpt_query *q);
/*
 * Move to next row in the query resultset
 */
bool ocrpt_navigate_next(opencreport *o, ocrpt_query *q);

#if 0
/*
 * Create a report part
 */
ocrpt_report_part *ocrpt_create_report_part(void);
/*
 * Add report header to report part
 */
ocrpt_report_part_set_report_header(ocrpt_report_part *part, ...);
#endif
/*
 * Add XML report description from a file
 */
int ocrpt_add_report(opencreport *o, const char *filename);
/*
 * Add XML report description using a memory buffer (zero-terminated string)
 */
int ocrpt_add_report_from_buffer(opencreport *o, const char *buffer);
/*
 * Parse the reports added up to this point.
 * Allow no deviation from the OpenCReports DTD.
 */
int ocrpt_parse(opencreport *o);
/*
 * Parse the reports added up to this point.
 * Optionally allow deviation from the DTD.
 */
int ocrpt_parse2(opencreport *o, int allow_bad_xml);
/*
 * Execute the reports added up to this point.
 * Parse any unparsed report XMLs with non-strict checks.
 */
int ocrpt_execute(opencreport *o);

#endif /* _OPENCREPORT_H_ */
