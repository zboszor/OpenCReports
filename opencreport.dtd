<!ELEMENT OpenCReport (Datasources?,Queries?,(Part|Report)*)>
<!ELEMENT Datasources (Datasource+)>
<!ELEMENT Datasource EMPTY>
<!ATTLIST Datasource
	name CDATA #REQUIRED
	type CDATA #REQUIRED
	host CDATA #IMPLIED
	unix_socket CDATA #IMPLIED
	port CDATA #IMPLIED
	dbname CDATA #IMPLIED
	user CDATA #IMPLIED
	password CDATA #IMPLIED
	connstr CDATA #IMPLIED
	optionfile CDATA #IMPLIED
	group CDATA #IMPLIED >
<!ELEMENT Queries (Query)+>
<!ELEMENT Query (#PCDATA)>
<!ATTLIST Query
	name CDATA #REQUIRED
	datasource CDATA #REQUIRED
	value CDATA #IMPLIED
	rows CDATA #IMPLIED
	cols CDATA #IMPLIED
	coltypes CDATA #IMPLIED
	follower_for CDATA #IMPLIED
	follower_expr CDATA #IMPLIED >
<!ELEMENT Part (PageHeader|PageFooter|ReportHeader|ReportFooter|pr)*>
<!ATTLIST Part
	layout CDATA #IMPLIED
	pagesAcross CDATA #IMPLIED
	pages_across CDATA #IMPLIED
	fontSize CDATA #IMPLIED
	font_size CDATA #IMPLIED
	orientation CDATA #IMPLIED
	topMargin CDATA #IMPLIED
	top_margin CDATA #IMPLIED
	leftMargin CDATA #IMPLIED
	left_margin CDATA #IMPLIED
	bottomMargin CDATA #IMPLIED
	bottom_margin CDATA #IMPLIED
	paperType CDATA #IMPLIED
	paper_type CDATA #IMPLIED
	iterations CDATA #IMPLIED
	suppressPageHeaderFirstPage CDATA #IMPLIED >
<!ELEMENT pr (pd)+>
<!ATTLIST pr
	layout CDATA #IMPLIED
	newpage CDATA #IMPLIED >
<!ELEMENT pd (Report*)>
<!ATTLIST pd
	width CDATA #IMPLIED
	height CDATA #IMPLIED
	border_width CDATA #IMPLIED
	border_color CDATA #IMPLIED >
<!ELEMENT Report ((Alternate|NoData)|(Breaks?,Variables?)|PageHeader|PageFooter|ReportHeader|ReportFooter|(Detail|Graph|Chart)*|xi:include)* >
<!ATTLIST Report
	pagesAcross CDATA #IMPLIED
	pages_across CDATA #IMPLIED
	fontSize CDATA #IMPLIED
	font_size CDATA #IMPLIED
	orientation CDATA #IMPLIED
	topMargin CDATA #IMPLIED
	top_margin CDATA #IMPLIED
	leftMargin CDATA #IMPLIED
	left_margin CDATA #IMPLIED
	bottomMargin CDATA #IMPLIED
	bottom_margin CDATA #IMPLIED
	paperType CDATA #IMPLIED
	paper_type CDATA #IMPLIED
	iterations CDATA #IMPLIED
	suppressPageHeaderFirstPage CDATA #IMPLIED

	query CDATA #IMPLIED
	xmlns:xi CDATA #IMPLIED
	suppress CDATA #IMPLIED
	uniquerow CDATA #IMPLIED
	detail_columns CDATA #IMPLIED
	column_pad CDATA #IMPLIED >
<!ELEMENT xi:include EMPTY>
<!ATTLIST xi:include
	href CDATA #REQUIRED >
<!ELEMENT Alternate (NoData) >
<!ELEMENT NoData (Output) >
<!ELEMENT ReportHeader (Output)* >
<!ELEMENT PageHeader (Output)* >
<!ELEMENT ReportFooter (Output)* >
<!ELEMENT PageFooter (Output)* >
<!ELEMENT Output (Line | HorizontalLine | Image)* >
<!ELEMENT Line (field | literal)* >
<!ATTLIST Line
	fontSize CDATA #IMPLIED
	bold CDATA #IMPLIED
	suppress CDATA #IMPLIED
	color CDATA #IMPLIED
	bgcolor CDATA #IMPLIED >
<!ELEMENT HorizontalLine EMPTY>
<!ATTLIST HorizontalLine
	size CDATA #IMPLIED
	indent CDATA #IMPLIED
	length CDATA #IMPLIED
	fontSize CDATA #IMPLIED
	suppress CDATA #IMPLIED
	bgcolor CDATA #IMPLIED >
<!ELEMENT Image EMPTY >
<!ATTLIST Image
	value CDATA #REQUIRED
	type CDATA #REQUIRED
	width CDATA #REQUIRED
	height CDATA #REQUIRED >
<!ELEMENT field EMPTY>
<!ATTLIST field
	value CDATA #REQUIRED
	delayed CDATA #IMPLIED
	precalculate CDATA #IMPLIED
	width CDATA #IMPLIED
	align CDATA #IMPLIED
	format CDATA #IMPLIED
	bgcolor CDATA #IMPLIED
	color CDATA #IMPLIED
	link CDATA #IMPLIED
	maxlines CDATA #IMPLIED
	col CDATA #IMPLIED >
<!ELEMENT literal (#PCDATA) >
<!ATTLIST literal
	align CDATA #IMPLIED
	bgcolor CDATA #IMPLIED
	color CDATA #IMPLIED
	width CDATA #IMPLIED
	bold CDATA #IMPLIED
	italics CDATA #IMPLIED
	link CDATA #IMPLIED
	translate CDATA #IMPLIED
	col CDATA #IMPLIED >
<!ELEMENT Detail (FieldHeaders?, FieldDetails?) >
<!ELEMENT FieldHeaders (Output)* >
<!ELEMENT FieldDetails (Output)* >
<!ELEMENT Graph (Plot+) >
<!ATTLIST Graph
	name CDATA #IMPLIED
	type CDATA #REQUIRED
	subtype CDATA #REQUIRED
	width CDATA #IMPLIED
	height CDATA #IMPLIED
	bold_titles CDATA #IMPLIED
	title CDATA #IMPLIED
	legend_bg_color CDATA #IMPLIED
	legend_orientation CDATA #IMPLIED
	draw_x_line CDATA #IMPLIED
	draw_y_line CDATA #IMPLIED
	grid_color CDATA #IMPLIED
	x_axis_title CDATA #IMPLIED
	y_axis_title CDATA #IMPLIED
	y_axis_mod CDATA #IMPLIED
	y_axis_title_right CDATA #IMPLIED
	y_axis_decimals CDATA #IMPLIED
	y_axis_decimals_right CDATA #IMPLIED >
<!ELEMENT Plot EMPTY>
<!ATTLIST Plot
	axis CDATA #IMPLIED
	field CDATA #IMPLIED
	label CDATA #IMPLIED
	side CDATA #IMPLIED
	disabled CDATA #IMPLIED
	color CDATA #IMPLIED >
<!ELEMENT Breaks (Break)* >
<!ELEMENT Break (BreakHeader | BreakFooter | BreakFields)* >
<!ATTLIST Break
	name CDATA #REQUIRED
	headernewpage CDATA #IMPLIED
	suppressblank CDATA #IMPLIED
	newpage CDATA #IMPLIED >
<!ELEMENT BreakHeader (Output)* >
<!ELEMENT BreakFooter (Output)* >
<!ELEMENT BreakFields (BreakField)* >
<!ELEMENT BreakField EMPTY>
<!ATTLIST BreakField
	value CDATA #REQUIRED >
<!ELEMENT Variables (Variable)* >
<!ELEMENT Variable EMPTY >
<!ATTLIST Variable
	name CDATA #REQUIRED
	value CDATA #REQUIRED
	type (expression|count|sum|average|lowest|highest) "sum"
	resetonbreak CDATA #IMPLIED
	precalculate CDATA #IMPLIED
	delayed CDATA #IMPLIED >
