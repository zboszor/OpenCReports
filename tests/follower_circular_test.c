/*
 * OpenCReports test
 * Copyright (C) 2019-2023 Zoltán Böszörményi <zboszor@gmail.com>
 * See COPYING.LGPLv3 in the toplevel directory.
 */

#include <stdio.h>

#include <opencreport.h>

#define ROWS 3
#define COLS 5
static const char *array[ROWS + 1][COLS] = {
	{ "id", "name", "property", "age", "adult" },
	{ "1", "Fred Flintstone", "strong", "31", "yes" },
	{ "2", "Wilma Flintstone", "charming", "28", "yes" },
	{ "3", "Pebbles Flintstone", "young", "5e-1", "no" }
};

static const int32_t coltypes[COLS] = {
	OCRPT_RESULT_NUMBER, OCRPT_RESULT_STRING, OCRPT_RESULT_STRING, OCRPT_RESULT_NUMBER, OCRPT_RESULT_NUMBER
};

#define ROWS1 2
static const char *array2[ROWS1 + 1][COLS] = {
	{ "id", "name", "property", "age", "adult" },
	{ "2", "Betty Rubble", "beautiful", "27", "yes" },
	{ "1", "Barney Rubble", "small", "29", "yes" },
};

int main(void) {
	opencreport *o = ocrpt_init();
	ocrpt_datasource *ds = ocrpt_datasource_add_array(o, "array");
	ocrpt_query *a, *b, *c, *d;
	ocrpt_expr *match_ab, *match_bc, *match_ac, *match_cd;
	bool retval;

	printf("added query a\n\n");
	a = ocrpt_query_add_array(ds, "a", (const char **)array, ROWS, COLS, coltypes, COLS);
	printf("added query b\n\n");
	b = ocrpt_query_add_array(ds, "b", (const char **)array2, ROWS1, COLS, coltypes, COLS);

	printf("adding N:1 follower a -> b, should succeed\n");
	match_ab = ocrpt_expr_parse(o, "a.id = b.id", NULL);
	retval = ocrpt_query_add_follower_n_to_1(a, b, match_ab);
	printf("added N:1 follower a -> b, retval %d\n\n", retval);

	printf("adding N:1 follower a -> b (duplicate), should fail\n");
	match_ab = ocrpt_expr_parse(o, "a.id = b.id", NULL);
	retval = ocrpt_query_add_follower_n_to_1(a, b, match_ab);
	printf("added N:1 follower a -> b, retval %d\n\n", retval);

	printf("adding follower a -> b (N:1 exists), should fail\n");
	retval = ocrpt_query_add_follower(a, b);
	printf("added follower a -> b, retval %d\n\n", retval);

	printf("adding N:1 follower b -> a (reverse N:1 exists), should fail\n");
	match_ab = ocrpt_expr_parse(o, "a.id = b.id", NULL);
	retval = ocrpt_query_add_follower_n_to_1(b, a, match_ab);
	printf("added N:1 follower b -> a, retval %d\n\n", retval);

	printf("adding follower b -> a (reverse N:1 exists), should fail\n");
	retval = ocrpt_query_add_follower(b, a);
	printf("added follower b -> a, retval %d\n\n", retval);

	c = NULL;

	printf("adding N:1 follower b -> c (query c does not exist), should fail\n");
	match_bc = ocrpt_expr_parse(o, "b.id = c.id", NULL);
	retval = ocrpt_query_add_follower_n_to_1(b, c, match_bc);
	printf("added N:1 follower b -> c, retval %d\n\n", retval);

	printf("adding follower b -> c (query c does not exist), should fail\n");
	retval = ocrpt_query_add_follower(b, c);
	printf("added follower b -> c, retval %d\n\n", retval);

	c = ocrpt_query_add_array(ds, "c", (const char **)array2, ROWS1, COLS, coltypes, COLS);
	printf("added query c\n\n");

	printf("adding N:1 follower b -> c, should succeed\n");
	match_bc = ocrpt_expr_parse(o, "b.id = c.id", NULL);
	retval = ocrpt_query_add_follower_n_to_1(b, c, match_bc);
	printf("added N:1 follower b -> c, retval %d\n\n", retval);

	printf("adding follower b -> c (N:1 follower exists), should fail\n");
	retval = ocrpt_query_add_follower(b, c);
	printf("added follower b -> c, retval %d\n\n", retval);

	printf("adding N:1 follower c -> a (circular followers), should fail\n");
	match_ac = ocrpt_expr_parse(o, "a.id = c.id", NULL);
	retval = ocrpt_query_add_follower_n_to_1(c, a, match_ac);
	printf("added N:1 follower c -> a, retval %d\n\n", retval);

	printf("adding follower c -> a (a->b->c exists), should fail\n");
	retval = ocrpt_query_add_follower(c, a);
	printf("added N:1 follower c -> a, retval %d\n\n", retval);

	printf("adding N:1 follower a -> c (c would be a follower on two paths), should fail\n");
	match_ac = ocrpt_expr_parse(o, "a.id = c.id", NULL);
	retval = ocrpt_query_add_follower_n_to_1(a, c, match_ac);
	printf("added N:1 follower a -> c, retval %d\n\n", retval);

	printf("adding follower a -> c (c would be a follower on two paths), should fail\n");
	retval = ocrpt_query_add_follower(a, c);
	printf("added follower a -> c, retval %d\n\n", retval);

	d = ocrpt_query_add_array(ds, "d", (const char **)array2, ROWS1, COLS, coltypes, COLS);
	printf("added query d\n\n");

	printf("adding follower c -> d, should succeed\n");
	match_cd = ocrpt_expr_parse(o, "c.id = d.id", NULL);
	retval = ocrpt_query_add_follower_n_to_1(c, d, match_cd);
	printf("added follower c -> d, retval %d\n\n", retval);

	printf("adding follower d -> a, should fail\n");
	match_cd = ocrpt_expr_parse(o, "c.id = d.id", NULL);
	retval = ocrpt_query_add_follower_n_to_1(d, a, match_cd);
	printf("added follower d -> a, retval %d\n\n", retval);

	ocrpt_free(o);

	return 0;
}
