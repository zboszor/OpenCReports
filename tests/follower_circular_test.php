<?php
/*
 * OpenCReports test
 * Copyright (C) 2019-2024 Zoltán Böszörményi <zboszor@gmail.com>
 * See COPYING.LGPLv3 in the toplevel directory.
 */

$array = [
	[ "id", "name", "property", "age", "adult" ],
	[ "1", "Fred Flintstone", "strong", "31", "yes" ],
	[ "2", "Wilma Flintstone", "charming", "28", "yes" ],
	[ "3", "Pebbles Flintstone", "young", "5e-1", "no" ]
];

$coltypes = [
	OpenCReport::RESULT_NUMBER, OpenCReport::RESULT_STRING, OpenCReport::RESULT_STRING, OpenCReport::RESULT_NUMBER, OpenCReport::RESULT_NUMBER
];

$array2 = [
	[ "id", "name", "property", "age", "adult" ],
	[ "2", "Betty Rubble", "beautiful", "27", "yes" ],
	[ "1", "Barney Rubble", "small", "29", "yes" ],
];

$o = new OpenCReport();
$ds = $o->datasource_add("array", "array");

echo "added query a" . PHP_EOL . PHP_EOL;
$a = $ds->query_add("a", "array", "coltypes");
echo "added query b" . PHP_EOL . PHP_EOL;
$b = $ds->query_add("b", "array2", "coltypes");

echo "adding N:1 follower a -> b, should succeed" . PHP_EOL;
$match_ab = $o->expr_parse("a.id = b.id");
$retval = $a->add_follower_n_to_1($b, $match_ab);
echo "added N:1 follower a -> b, retval " . (int)$retval . PHP_EOL . PHP_EOL;

echo "adding N:1 follower a -> b (duplicate), should fail" . PHP_EOL;
$match_ab = $o->expr_parse("a.id = b.id");
$retval = $a->add_follower_n_to_1($b, $match_ab);
echo "added N:1 follower a -> b, retval " . (int)$retval . PHP_EOL . PHP_EOL;

echo "adding follower a -> b (N:1 exists), should fail" . PHP_EOL;
$retval = $a->add_follower($b);
echo "added follower a -> b, retval " . (int)$retval . PHP_EOL . PHP_EOL;

echo "adding N:1 follower b -> a (reverse N:1 exists), should fail" . PHP_EOL;
$match_ab = $o->expr_parse("a.id = b.id");
$retval = $b->add_follower_n_to_1($a, $match_ab);
echo "added N:1 follower b -> a, retval " . (int)$retval . PHP_EOL . PHP_EOL;

echo "adding follower b -> a (reverse N:1 exists), should fail" . PHP_EOL;
$retval = $b->add_follower($a);
echo "added follower b -> a, retval " . (int)$retval . PHP_EOL . PHP_EOL;

unset($c);

echo "adding N:1 follower b -> c (query c does not exist), should fail" . PHP_EOL;
$match_bc = $o->expr_parse("b.id = c.id");
try {
	$retval = $b->add_follower_n_to_1($c, $match_bc);
} catch (TypeError $e) {
	$retval = false;
}
echo "added N:1 follower b -> c, retval " . (int)$retval . PHP_EOL . PHP_EOL;

echo "adding follower b -> c (query c does not exist), should fail" . PHP_EOL;
try {
	$retval = $b->add_follower($c);
} catch (TypeError $e) {
	$retval = false;
}
echo "added follower b -> c, retval " . (int)$retval . PHP_EOL . PHP_EOL;

$c = $ds->query_add("c", "array2", "coltypes");
echo "added query c" . PHP_EOL . PHP_EOL;

echo "adding N:1 follower b -> c, should succeed" . PHP_EOL;
$match_bc = $o->expr_parse("b.id = c.id");
$retval = $b->add_follower_n_to_1($c, $match_bc);
echo "added N:1 follower b -> c, retval " . (int)$retval . PHP_EOL . PHP_EOL;

echo "adding follower b -> c (N:1 follower exists), should fail" . PHP_EOL;
$retval = $b->add_follower($c);
echo "added follower b -> c, retval " . (int)$retval . PHP_EOL . PHP_EOL;

echo "adding N:1 follower c -> a (circular followers), should fail" . PHP_EOL;
$match_ac = $o->expr_parse("a.id = c.id");
$retval = $c->add_follower_n_to_1($a, $match_ac);
echo "added N:1 follower c -> a, retval " . (int)$retval . PHP_EOL . PHP_EOL;

echo "adding follower c -> a (a->b->c exists), should fail" . PHP_EOL;
$retval = $c->add_follower($a);
echo "added N:1 follower c -> a, retval " . (int)$retval . PHP_EOL . PHP_EOL;

echo "adding N:1 follower a -> c (c would be a follower on two paths), should fail" . PHP_EOL;
$match_ac = $o->expr_parse("a.id = c.id");
$retval = $a->add_follower_n_to_1($c, $match_ac);
echo "added N:1 follower a -> c, retval " . (int)$retval . PHP_EOL . PHP_EOL;

echo "adding follower a -> c (c would be a follower on two paths), should fail" . PHP_EOL;
$retval = $a->add_follower($c);
echo "added follower a -> c, retval " . (int)$retval . PHP_EOL . PHP_EOL;

$d = $ds->query_add("d", "array2", "coltypes");
echo "added query d" . PHP_EOL . PHP_EOL;

echo "adding follower c -> d, should succeed" . PHP_EOL;
$match_cd = $o->expr_parse("c.id = d.id");
$retval = $c->add_follower_n_to_1($d, $match_cd);
echo "added follower c -> d, retval " . (int)$retval . PHP_EOL . PHP_EOL;

echo "adding follower d -> a, should fail" . PHP_EOL;
$match_cd = $o->expr_parse("c.id = d.id");
$retval = $d->add_follower_n_to_1($a, $match_cd);
echo "added follower d -> a, retval " . (int)$retval . PHP_EOL . PHP_EOL;
