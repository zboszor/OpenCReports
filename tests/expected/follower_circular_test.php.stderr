PHP Warning:  OpenCReport\Query::add_follower_n_to_1(): array:b follower would create a circular graph for array:a leader
 in /home/zozo/git/OpenCReports/tests/follower_circular_test.php on line 40
PHP Warning:  OpenCReport\Query::add_follower(): array:b follower would create a circular graph for array:a leader
 in /home/zozo/git/OpenCReports/tests/follower_circular_test.php on line 44
PHP Warning:  OpenCReport\Query::add_follower_n_to_1(): array:a follower would create a circular graph for array:b leader
 in /home/zozo/git/OpenCReports/tests/follower_circular_test.php on line 49
PHP Warning:  OpenCReport\Query::add_follower(): array:a follower would create a circular graph for array:b leader
 in /home/zozo/git/OpenCReports/tests/follower_circular_test.php on line 53
PHP Warning:  Undefined variable $c in /home/zozo/git/OpenCReports/tests/follower_circular_test.php on line 61
PHP Warning:  Undefined variable $c in /home/zozo/git/OpenCReports/tests/follower_circular_test.php on line 69
PHP Warning:  OpenCReport\Query::add_follower(): array:c follower would create a circular graph for array:b leader
 in /home/zozo/git/OpenCReports/tests/follower_circular_test.php on line 84
PHP Warning:  OpenCReport\Query::add_follower_n_to_1(): array:a follower would create a circular graph for array:c leader
 in /home/zozo/git/OpenCReports/tests/follower_circular_test.php on line 89
PHP Warning:  OpenCReport\Query::add_follower(): array:a follower would create a circular graph for array:c leader
 in /home/zozo/git/OpenCReports/tests/follower_circular_test.php on line 93
PHP Warning:  OpenCReport\Query::add_follower_n_to_1(): array:c follower would create a circular graph for array:a leader
 in /home/zozo/git/OpenCReports/tests/follower_circular_test.php on line 98
PHP Warning:  OpenCReport\Query::add_follower(): array:c follower would create a circular graph for array:a leader
 in /home/zozo/git/OpenCReports/tests/follower_circular_test.php on line 102
PHP Warning:  OpenCReport\Query::add_follower_n_to_1(): array:a follower would create a circular graph for array:d leader
 in /home/zozo/git/OpenCReports/tests/follower_circular_test.php on line 115
